scm/github/app_install.go

```diff
@@ -0,0 +1,156 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package github
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"gorm.io/gorm"
+
+	"github.com/go-vela/server/api/types"
+	"github.com/go-vela/server/constants"
+	"github.com/go-vela/server/database"
+	"github.com/go-vela/server/internal"
+)
+
+// ProcessInstallation takes a GitHub installation and processes the changes.
+func (c *client) ProcessInstallation(ctx context.Context, _ *http.Request, webhook *internal.Webhook, db database.Interface) error {
+	c.Logger.Tracef("processing GitHub App installation")
+
+	errs := []string{}
+
+	// set install_id for repos added to the installation
+	for _, repo := range webhook.Installation.RepositoriesAdded {
+		r, err := db.GetRepoForOrg(ctx, webhook.Installation.Org, repo)
+		if err != nil {
+			if !errors.Is(err, gorm.ErrRecordNotFound) {
+				errs = append(errs, fmt.Sprintf("%s:%s", repo, err.Error()))
+			}
+
+			// skip repos that dont exist in vela
+			continue
+		}
+
+		err = updateRepoInstallationID(ctx, webhook, r, db, webhook.Installation.ID)
+		if err != nil {
+			errs = append(errs, fmt.Sprintf("%s:%s", repo, err.Error()))
+		}
+	}
+
+	// set install_id for repos removed from the installation
+	for _, repo := range webhook.Installation.RepositoriesRemoved {
+		r, err := db.GetRepoForOrg(ctx, webhook.Installation.Org, repo)
+		if err != nil {
+			if !errors.Is(err, gorm.ErrRecordNotFound) {
+				errs = append(errs, fmt.Sprintf("%s:%s", repo, err.Error()))
+			}
+
+			// skip repos that dont exist in vela
+			continue
+		}
+
+		err = updateRepoInstallationID(ctx, webhook, r, db, 0)
+		if err != nil {
+			errs = append(errs, fmt.Sprintf("%s:%s", repo, err.Error()))
+		}
+	}
+
+	// combine all errors
+	if len(errs) > 0 {
+		return errors.New(strings.Join(errs, ", "))
+	}
+
+	return nil
+}
+
+// updateRepoInstallationID updates the installation ID for a repo.
+func updateRepoInstallationID(ctx context.Context, webhook *internal.Webhook, r *types.Repo, db database.Interface, installID int64) error {
+	r.SetInstallID(installID)
+
+	h := new(types.Hook)
+	h.SetNumber(webhook.Hook.GetNumber())
+	h.SetSourceID(webhook.Hook.GetSourceID())
+	h.SetWebhookID(webhook.Hook.GetWebhookID())
+	h.SetCreated(webhook.Hook.GetCreated())
+	h.SetHost(webhook.Hook.GetHost())
+	h.SetEvent(constants.EventInstallation)
+	h.SetStatus(webhook.Hook.GetStatus())
```

Comment:
the webhook.Hook should already have all this right? Is there a reason we create a new one here?

Reply:
yeah its a deep copy of the base hook for each installed repo since thats a one-to-many for install-to-repohooks.
we could re-use webhook.Hook but felt cleaner to do a deep copy and not mutate the input pointer.